core/common/src/main/java/alluxio/conf/AlluxioProperties.java

```diff
@@ -57,6 +59,9 @@
   /** Map of property sources. */
   private final ConcurrentHashMap<PropertyKey, Source> mSources = new ConcurrentHashMap<>();
 
+  /** A value representing the running hash code for this set of properties. */
+  private AtomicInteger mPrecomputedHash = new AtomicInteger(0);
```

Comment:
Can we implement `hashCode`/`equals` on immutable configuration objects instead? Caching the hash code would be much easier there since it doesn't change. Also, are we certain that it's necessary to perform caching? It seems like we would only call `hashCode` a single time when putting it in the cache.

Reply:
The only thing that's immutable is the `AlluxioConfiguration` interface. You can't enforce implementations to implement the `hashCode` method. InstancedConfiguration. I chose to implement it on `AlluxioProperties` rather than `InstancedConfiguration` The idea around pre-computing it is so that when a user creates a configuration and needs to retrieve a client that the hashCode would need to be computed for every call to `Filesystem.get`. Microbench creates a huge number clients and we would like to minimize the performance impact of doing the lookup.